localgovdrupal/localgov_forms
# OS Places geocoder

## Summary of code changes
- ~~[PHP Geocoder plugin](https://geocoder-php.org/) for the [OS Places API](https://osdatahub.os.uk/docs/places/overview).  This is the plugin [File](https://github.com/localgovdrupal/localgov_forms/blob/feature/os-places-geocoder/src/Geocoder/Provider/LocalgovOsPlacesGeocoder.php).~~
- ~~Drupal Geocoder [plugin](https://github.com/localgovdrupal/localgov_forms/blob/feature/os-places-geocoder/src/Plugin/Geocoder/Provider/LocalgovOsPlacesGeocoder.php) for the OS Places API.  This one is a Drupal specific wrapper for the above plugin.~~
- Further changes to the LocalGovAddress class to make room for OS specific fields (e.g. Building number, Organisation name, etc.)
- Improvements to how we prepare address line 1 and address line 2.

## Test steps
- Testing this could be difficult as the OS Places API doesn't seem to have a [free plan](https://osdatahub.os.uk/plans).  But I could be wrong.
- Assuming you are in possession of an API key, Add the *Localgov OS Places* Geocoder plugin from /admin/config/system/geocoder/geocoder-provider
- Add the Localgov address lookup Webform element to any Webform. While configuring this Webform element, select the *Localgov OS Places* Geocoder plugin.
- Search for any postcode.  Special ones like XM4 5HQ won't work as the OS Places API doesn't recognise these.  But others should be fine.

## Notes
- The OS Places API is returning [All numeric grid references](https://en.wikipedia.org/wiki/Ordnance_Survey_National_Grid#All-numeric_grid_references) instead of latitude/longitude.  So the latitude and longitude values for each address entry are empty.  Easting and Northing Grid reference values are available as separate fields.  Rich has pointed out that there are conversion algorithms available.  And they are [indeed available](https://scipython.com/book2/chapter-2-the-core-python-language-i/additional-problems/converting-between-an-os-grid-reference-and-longitudelatitude/). We will have to implement that in a future ticket.  This will allow us to populate the latitude and longitude values.
- The OS Places API is returning building numbers containing alphabets (e.g. 30B, 9B-5C) as **Building names** instead of **Building numbers**.  This breaks the address formatting as I have placed Building names on the first address line and Building numbers on the second.  This may need some further processing.  I am thinking of leaving this for a later ticket.  I hope that's acceptable.
- Ekes wanted the Geocoder plugin to be a stand-alone composer package and the Drupal Geocoder plugin to be part of the localgov_geo module.  He has also just confirmed this in a comment below.  ~~I will take steps to relocate the relevant files to those repos in consultation with Ekes.~~ This is done now.

@andybroomfield @andrew-wallis @ekes @finnlewis @MattOz-CDS @richardaclarke @tom-steel
